qupath-core-processing/src/main/java/qupath/opencv/tools/OpenCVTools.java

```diff
@@ -2651,6 +2651,70 @@ public static List<IndexedPixel> getMaskedPixels(Mat mat, Mat mask) {
 		}
 		return pixels;
 	}
+
+	/**
+	 * <p>Extract pixels from an image using a mask, limited to the specified channel index.</p>
+	 * <p>The mask can have one or multiple channels:</p>
+	 * <ul>
+	 *     <li>If the mask has only one channel, this channel is used.</li>
+	 *     <li>If the mask is multichannel, only the specified channel is used.</li>
+	 * </ul>
+	 * <p>The input and the mask must have the same dimensions.</p>
+	 *
+	 * @param input  the image from where the pixels should be extracted
+	 * @param mask  the mask to apply to the input
+	 * @param channel  the channel index (zero-based) to consider
+	 * @return a float array containing the values of the input filtered by the mask
+	 * @throws IllegalArgumentException when the input and the mask don't have the same dimensions
+	 */
+	public static float[] extractMaskedPixels(Mat input, Mat mask, int channel) {
+		Mat inputChannel = new Mat();
+		opencv_core.extractChannel(input, inputChannel, channel);
+
+		Mat maskChannel = new Mat();
+		if (mask.channels() == 1) {
+			maskChannel = mask;
+		} else {
+			opencv_core.extractChannel(mask, maskChannel, channel);
+		}
+
+		if (inputChannel.rows() != maskChannel.rows()) {
+			throw new IllegalArgumentException("The input and the mask don't have the same number of rows");
+		}
+		if (inputChannel.cols() != maskChannel.cols()) {
+			throw new IllegalArgumentException("The input and the mask don't have the same number of columns");
+		}
+
+		int height = inputChannel.rows();
+		int width = inputChannel.cols();
+		List<Float> outputList = new ArrayList<>();
```

Comment:
This is likely to be a performance bottleneck, because `float` will need to be boxed to `Float`.

I guess the reason to use a list is that you can't know in advance how many pixels you'll need unless you iterate through the mask first, which would also have some performance hit.

My guess is that a good balance of efficiency would be to preallocate a `float[]` array capable of storing all the image pixels, and then populating it with unmasked pixels during iteration - storing an index, and skipping masked-out pixels.

And the end, you can check if the index is equal to the array length - if so, just return the array (i.e. the mask includes everything). If not, a single call to [`Arrays.copyOf(float[], int)`](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/util/Arrays.html#copyOf(float[],int)) would return an array of the correct length.